commit d9848bb2a0e1ad26d72133570e71d034055c5760
Author: bertin <macportsuser@macports.org>
Date:   Thu May 22 23:41:31 2025 +0200

    /opt/local/site-ports/sysutils/MacPorts/files/patch-cxxstdlib-treatment.diff

diff --git a/src/port1.0/portconfigure.tcl b/src/port1.0/portconfigure.tcl
index 680304412..6fef794b6 100644
--- a/src/port1.0/portconfigure.tcl
+++ b/src/port1.0/portconfigure.tcl
@@ -99,6 +99,10 @@ proc portconfigure::should_add_stdlib {} {
             if { ${gcc_ver} eq "devel" || ${gcc_ver} >= 10 } {
                 return 1
             }
+        } else {
+            # oh, just make sure that our caller adds -stdlib=libstdc++, so that
+            # GCC builds that use libc++ by default can be "toggled back"
+            return 1
         }
     }
     return 0
@@ -114,34 +118,49 @@ proc portconfigure::should_add_cxx_abi {} {
                   ${os.platform} eq "darwin" &&
                   ${configure.cxx_stdlib} eq "macports-libstdc++"}]
 }
+
 proc portconfigure::construct_cxxflags {flags} {
     if {[portconfigure::should_add_stdlib]} {
         global configure.cxx_stdlib
-        lappend flags -stdlib=${configure.cxx_stdlib}
+        if {[string match */g*-mp-* ${configure.cxx}]
+            && ${configure.cxx_stdlib} eq "macports-libstdc++"} {
+            lappend flags -stdlib=libstdc++
+        } else {
+            lappend flags -stdlib=${configure.cxx_stdlib}
+        }
     }
     if {[portconfigure::should_add_cxx_abi]} {
         lappend flags -D_GLIBCXX_USE_CXX11_ABI=0
     }
     return $flags
 }
+
 proc portconfigure::stdlib_trace {opt action args} {
     global $opt
     foreach flag [lsearch -all -inline [set $opt] -stdlib=*] {
         $opt-delete $flag
     }
-    foreach flag [lsearch -all -exact -inline [set $opt] -D_GLIBCXX_USE_CXX11_ABI=0] {
-        $opt-delete $flag
+    if {[option os.platform] eq "darwin"} {
+        foreach flag [lsearch -all -exact -inline [set $opt] -D_GLIBCXX_USE_CXX11_ABI=0] {
+            $opt-delete $flag
+        }
     }
     if {$action eq "read"} {
         if {[portconfigure::should_add_stdlib]} {
-            global configure.cxx_stdlib
-            $opt-append -stdlib=${configure.cxx_stdlib}
+            global configure.cxx_stdlib configure.cxx
+            if {[string match */g*-mp-* ${configure.cxx}]
+                && ${configure.cxx_stdlib} eq "macports-libstdc++"} {
+                $opt-append -stdlib=libstdc++
+            } else {
+                $opt-append -stdlib=${configure.cxx_stdlib}
+            }
         }
         if {[portconfigure::should_add_cxx_abi]} {
             $opt-append -D_GLIBCXX_USE_CXX11_ABI=0
         }
     }
 }
+
 # helper function to set configure.cxx_stdlib
 proc portconfigure::configure_get_cxx_stdlib {} {
     global cxx_stdlib compiler.cxx_standard
